examples/wave_sim/wave_sim.cc

```diff
@@ -63,23 +63,33 @@ void update(celerity::distr_queue& queue, celerity::buffer<float, 2> up, celerit
 	step<float, update_config, class update>(queue, up, u, dt, delta);
 }
 
+void stream_open(celerity::distr_queue& queue, size_t N, size_t num_samples, celerity::experimental::host_object<std::ofstream> os) {
+	queue.submit([=](celerity::handler& cgh) {
+		celerity::experimental::side_effect os_eff{os, cgh};
+		cgh.host_task(celerity::on_master_node, [=] {
+			os_eff->open("wave_sim_result.bin", std::ios_base::out | std::ios_base::binary);
+			const struct { uint64_t n, t; } header{N, num_samples};
```

Comment:
Hmm, I'm not sure what to check for here, the layout of the struct is well-defined. The one thing I could think of is integer endianness. That will be easy to check with C++20 `std::endian`, but until then, we'd have to build our own with `__BYTE_ORDER__` (GCC / Clang) or MSVC-specifics on Windows. @PeterTh I'm not fluent in Ruby, does the script assume a particular endianness in the input?

Reply:
Yes, in practice endianness is the only thing that could break, but in theory an architecture could have an alignment requirement greater 8 for `uint64_t`, which would mean that the struct would be padded and so no longer conforms to the layout requirement. However, this will never happen on any real architecture, but I still think it's a good idea to `static_assert` such things, even if it's just to document requirements (and it would protect against adding more members to the struct).

As for endianness, I don't think it's worth implementing compiler specific versions. But maybe we can have a commented out use of `std::endian` that we enable when we switch to C++20?

Reply:
> but in theory an architecture could have an alignment requirement greater 8 for uint64_t

That can't happen, because the standard guarantees that the size of a type is an integer multiple of its alignment (otherwise arrays could require padding and `sizeof(T[n]) == n * sizeof(T)` would not hold).

We can do the C++20 thing, but I wonder if it's worth it considering that this is example code.

Reply:
> That can't happen, because the standard guarantees that the size of a type is an integer multiple of its alignment (otherwise arrays could require padding and sizeof(T[n]) == n * sizeof(T) would not hold).

Right, of course. I guess `static_assert`ing would then just protect against adding another member to the struct, which in this case is not very likely to happen.

> We can do the C++20 thing, but I wonder if it's worth it considering that this is example code.

Probably not. I guess it's fine as-is.

Reply:
Then I'll leave it alone for now.